src/MPoly.jl

```diff
@@ -425,16 +425,19 @@ end
 
 iszero(x::MPolyRingElem{T}) where T <: RingElement = length(x) == 0
 
-function is_unit(a::MPolyRingElem{T}) where T <: RingElement
-   if is_constant(a)
-      return is_unit(leading_coefficient(a))
-   elseif is_domain_type(elem_type(coefficient_ring(a)))
-      return false
-   elseif length(a) == 1
-      return false
-   else
-      throw(NotImplementedError(:is_unit, a))
-   end
+function is_unit(f::T) where { T<:MPolyRingElem }
+  # for constant polynomials we delegate to the coefficient ring:
+  is_constant(f) && return is_unit(constant_coefficient(f))
+  # Here deg(f) > 0; over an integral domain, non-constant polynomials are never units:
+  is_domain_type(elem_type(coefficient_ring(f))) && return false
+  for (c, expv) in zip(coefficients(f), exponent_vectors(f))
+    if is_zero(expv)
+      is_unit(c) || return false
+    else
+      is_nilpotent(c) || return false;
```

Comment:
```suggestion
      is_nilpotent(c) || return false
```
Semicolons keep creeping in ;-).


Anyway, this looks really nice now, thanks :-)